py/parse: Recognise const int assignments with type hints.

agatti opened the pull request:
### Summary

This PR extends the parser's behaviour to also recognise constant integer assignments even when the assignment itself has an "int" type annotation.

Now declarations like "var: int = const(val)" can be treated as constants by the compiler and thus be folded in the rest of the program as it sees fit.  Before this change, that line would generate a variable creation and its value assignment, without any folding being performed.

This fixes #15608.

### Testing

This was tested using the same verification methodology described in #15608, and the Unix test suite was ran locally without errors with this change applied.

### Trade-offs and Alternatives

I've limited the type checking to `int` types, which I hope it's all that's needed.  It can probably be improved style-wise, but from what I reckon the logic *should* be correct - I haven't looked at the parser since the rv32 inline asm days.

agatti committed:
py/parse: Recognise const assignments with type hints.

This commit extends the parser's behaviour to also recognise constant
integer assignments even when the assignment itself has a type
annotation.

Now declarations like "var: int = const(val)" can be treated as
constants by the compiler and thus be folded in the rest of the program
as it sees fit.  Before this change, that line would generate a variable
creation and its value assignment, without any folding being performed.

This fixes #15608.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
Forgot to add tests... thank you, codecov!  I've also had the chance to reword the commit message as it wasn't entirely clear.  I'll push the lot once the current test run is over to avoid having to receive loads of cancelled run reports.

Anyway, I wonder how difficult it would be to have something like `mpy-tool.py`'s disassembler available to the tests' running environment.  This way the tests could also verify that, for a given bit of code, optimised bytecode was emitted instead of regular code sequences - that would make some sense for this specific PR at least.

agatti commented:
I've just tried, and it does indeed work - after all the type annotations are discarded anyway.  Also, I've simplified the code and turned out the final overhead is bigger, go figure.  I'll revert the PR to what it looked like then.

From an end user's point of view, having annotated variables exhibit the same behaviour as non-annotated ones is probably a reasonable expectation, isn't it?

dlech commented:
Yes, a reasonable expectation. And the compiler can be disabled on super-constrained systems. So the code size isn't as much of a concern as I initially made it to be.

agatti commented:
> Actually, const supports bool and tuple. And also now float! So the type annotation could be rather complicated if you have nested tuples etc.

Indeed.  Luckily I limited myself to just handle `int` types with this!  I've tried to extend the const condition check to cover both `MP_QSTR_int` and `MP_QSTR_float` as the checked annotation type but, unless I did the test incorrectly, it didn't seem to work as expected.

I can revisit this at another time though, so both `float` and `bool` can be handled this way as well.

> There are tests for generated bytecode, in particular one for const generation. See tests/cmdline/cmd_showbc_const.py.

Thanks, I didn't know this was possible.  Although what I had in mind was more or less a MicroPython implementation of CPython's `dis` module so unrelated future changes in the compiled output would not break the test unless they affect const loads.

dlech commented:
Why do we need to check the annotation value and limit it to only `int`? Can't we just remove `&& MP_PARSE_NODE_LEAF_ARG(((mp_parse_node_struct_t *)pn1)->nodes[0]) == MP_QSTR_int)` and basically ignore the annotation to allow all const expressions with annotations?

agatti commented:
Thanks for the test template.  The previous iteration duplicated the original tests because they only dealt with integers in the first place, so that would fit with the initial reduced scope of these changes.

Anyway, I've added a missing const-able type (`bytes`) to your suggestion and removed the rest of the duplicated code.

dpgeorge reviewed (approved):
Thanks for updating the test.
